src/v/ssx/include/ssx/sharded_int.h

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "base/seastarx.h"
+
+#include <seastar/core/smp.hh>
+
+#include <algorithm>
+#include <atomic>
+#include <memory>
+#include <type_traits>
+#include <vector>
+
+namespace ssx {
+
+/// sharded_int is an atomic primitive that has a local copy of an integer
+/// on each shard that can be independently updated and provides a helper for
+/// gathering the maximum of the per-shard values. The local copies do not have
+/// the same value, they are updated independently.
+template<typename T>
+requires std::is_integral_v<T>
+class sharded_int {
+public:
+    sharded_int(T value)
+      : _state(ss::smp::count, value) {}
+    ~sharded_int() noexcept = default;
+
+    sharded_int(sharded_int&& other) noexcept = default;
+    sharded_int& operator=(sharded_int&&) noexcept = default;
+
+    sharded_int(sharded_int const&) = delete;
+    sharded_int& operator=(sharded_int const&) = delete;
+
+    /// return the maximum value across the shards
+    ss::future<T> max() const {
+        std::vector<std::atomic<T>> state_copy(ss::smp::count);
+        co_await ss::smp::invoke_on_all([this, &state_copy]() {
+            state_copy[ss::this_shard_id()].store(
+              this->local(), std::memory_order_seq_cst);
+        });
+
+        T result;
+        for (const auto& x : state_copy) {
+            result = std::max(result, x.load(std::memory_order_seq_cst));
+        }
```

Comment:
Could use `std::reduce` here.

Reply:
Rewrote `gc()` using `map_reduce` to avoid blocking inside the `sharded_ptr::update`.